src/color-scale/color-scale.js

```diff
@@ -17,31 +20,253 @@ const Self = class ColorScale extends ColorElement {
 			slot: this.shadowRoot.querySelector("slot"),
 			swatches: this.shadowRoot.getElementById("swatches"),
 		};
+
+		this._slots = {
+			add_button: this.shadowRoot.querySelector("slot[name=add-button]"),
+		};
 	}
 
 	connectedCallback () {
 		super.connectedCallback?.();
+		this._el.swatches.addEventListener("labelchange", this, {capture: true});
 		this._el.swatches.addEventListener("colorchange", this, {capture: true});
+		this._el.swatches.addEventListener("click", this, {capture: true});
+		this._slots.add_button.addEventListener("click", this);
 	}
 
 	disconnectedCallback () {
 		this.#swatches = [];
+		this._el.swatches.removeEventListener("labelchange", this, {capture: true});
 		this._el.swatches.removeEventListener("colorchange", this, {capture: true});
+		this._el.swatches.removeEventListener("click", this, {capture: true});
+		this._slots.add_button.removeEventListener("click", this);
 	}
 
 	handleEvent (event) {
+		let source = event.target;
+
+		if (event.type === "click" && source === this._el.add_button || this._slots.add_button.assignedElements().includes(source)) {
+			this.addColor();
+			return;
+		}
+
+		if (event.type === "colorchange" && source.matches("color-swatch:not(.intermediate, .ignore-updates)")) {
+			this.updateColor(source);
+		}
+		else if (event.type === "labelchange" && source.matches("color-swatch:not(.intermediate, .ignore-updates)")) {
+			this.updateColorName(source);
+		}
+		else if (event.type === "click" && source.closest("button[part=delete-button]")) {
+			this.deleteColor(source.closest("color-swatch"));
+		}
+
 		this.dispatchEvent(new event.constructor(event.type, {...event}));
 	}
 
 	propChangedCallback ({name, prop, detail: change}) {
-		if (name === "computedColors") {
+		if (name === "computedColors" && !this.editable?.name && !this.editable?.color) {
 			// Re-render swatches
+			// Only if nothing is being edited, otherwise the input would be lost
+			// or, e.g., "red" would be converted to "rgb(100%, 0%, 0%)" right after the typing is done
+			this.render();
```

Comment:
I would suggest instead checking if it's focused (via `document.activeElement` IIRC). It's entirely possible that it's editable but nobody is interacting with it right now, in which case re-rendering would be fine, right?
Though re-rendering seems a very crude reaction to `computedColors` changing, when does this actually happen? Is it when the entire array if overwritten? In that case I guess it makes sense…

Reply:
> I would suggest instead checking if it's focused (via `document.activeElement` IIRC). It's entirely possible that it's editable but nobody is interacting with it right now, in which case re-rendering would be fine, right?

You are absolutely right. Thank you for the idea. It works brilliantly. 

> Though re-rendering seems a very crude reaction to `computedColors` changing, when does this actually happen? Is it when the entire array if overwritten?

It does happen when `this.colors` is overwritten; you are right.